src/planner.c

```diff
@@ -312,8 +336,55 @@ timescaledb_planner(Query *parse, int cursor_opts, ParamListInfo bound_params)
 
 	PG_TRY();
 	{
+		PreprocessQueryContext context = { 0 };
+		context.rootquery = parse;
 		if (ts_extension_is_loaded())
-			preprocess_query((Node *) parse, parse);
+		{
+			preprocess_query((Node *) parse, &context);
+
+			/*
+			 * Determine which type of fetcher to use. If set by GUC, use what
+			 * is set. If the GUC says 'auto', use the row-by-row fetcher if we
+			 * have at most one distributed table in the query. This enables
+			 * parallel plans on data nodes, which speeds up the query.
+			 * We can't use parallel plans with the cursor fetcher, because the
+			 * cursors don't support parallel execution. This is because a
+			 * cursor can be suspended at any time, then some arbitrary user
+			 * code can be executed, and then the cursor is resumed. The
+			 * parallel infrastructure doesn't have enough reentrability to
+			 * survive this.
+			 * We have to use a cursor fetcher when we have multiple distributed
+			 * tables, because we might first have to get some rows from one
+			 * table and then from another, without running either of them to
+			 * completion first. This happens e.g. when doing a join. If we had
+			 * a connection per table, we could avoid this requirement.
+			 *
+			 * Note that this function can be called recursively, e.g. when
+			 * trying to evaluate an SQL function at the planning stage. We must
+			 * only set/reset the fetcher type at the topmost level, that's why
+			 * we check it's not already set.
+			 */
+			if (ts_data_node_fetcher_scan_type == AutoFetcherType)
+			{
+				reset_fetcher_type = true;
+
+				if (ts_guc_remote_data_fetcher == AutoFetcherType)
+				{
+					if (context.num_distributed_tables >= 2)
```

Comment:
Does this cover the case when you have a subquery from the same table? E.g., something like:

```sql
SELECT * FROM myhypertable WHERE  device=(SELECT device FROM myhypertable LIMIT 1);
```

In other words, does this count as two distributed hypertables or just one?

Reply:
Two, because they might be scanned concurrently, no matter whether it's the same physical table or not. Added a comment for num_distributed_tables.